### PR TITLE
Orthogonal decomposition

### PR DESCRIPTION
**Pull request by eggrobin**
It looks like this speeds up `AnalyticalSeriesTest.CompactRepresentation` between 10% and 20%.

### CODE REVIEW

File: numerics/poisson_series_basis_body.hpp [hunk 1]
```diff
@@ -92,6 +101,18 @@ std::array<Series, dimension * (degree + 1)> PoissonSeriesBasisGenerator<
               Aperiodic(origin)...};
 }
 
+template<typename Series, int dimension, int degree, std::size_t... indices>
+inline std::array<PoissonSeriesSubspace, dimension*(degree + 1)>
+PoissonSeriesBasisGenerator<
+    Series,
+    dimension,
+    degree,
```
Comment (pleroy) [inline numerics/poisson_series_basis_body.hpp:109]: Put all 3 parameters on a single line, similar to line 98.  In general, `clang-format` doesn't do a great job with these templates...

File: numerics/poisson_series_basis_body.hpp [hunk 2]
```diff
@@ -137,6 +158,26 @@ PoissonSeriesBasisGenerator<
   return all_series;
 }
 
+template<typename Series, int dimension, int degree, std::size_t... indices>
+inline std::array<PoissonSeriesSubspace, 2 * dimension * (degree + 1)>
+PoissonSeriesBasisGenerator<
+    Series,
+    dimension,
+    degree,
```
Comment (pleroy) [inline numerics/poisson_series_basis_body.hpp:166]: Same.

**Review by pleroy — COMMENTED** [review]
Strangely, the unit tests fail with major accuracy changes.  This must be investigated.

File: numerics/frequency_analysis_body.hpp [hunk 3]
```diff
@@ -112,20 +112,35 @@ IncrementalProjection(Function const& function,
   CHECK(ω.has_value());
 
   std::vector<Series> basis;
+  // The Poisson series basis[k] belongs to the subspace basis_subspaces[k];
+  // this remains true after orthonormalization, i.e., q[k] belongs to the
+  // subspace basis_subspaces[k] below.
+  std::vector<PoissonSeriesSubspace> basis_subspaces;
 
   int basis_size;
   if (ω.value() == AngularFrequency{}) {
```
Comment (pleroy) [inline numerics/frequency_analysis_body.hpp:121]: Please put a `TODO(phl)` to refactor this if statement and the one near line 170 in a separate function.

File: numerics/poisson_series_basis.hpp [hunk 4]
```diff
@@ -17,6 +17,41 @@ using geometry::Hilbert;
...
+// The type |PoissonSeriesSubspace| defines an orthogonal decomposition of the
+// space of Poisson series, i.e., The space of Poisson series is the orthogonal
+// sum of all values of |PoissonSeriesSubspace|.
+class PoissonSeriesSubspace {
+ public:
+  PoissonSeriesSubspace(PoissonSeriesSubspace const&) = default;
+  PoissonSeriesSubspace(PoissonSeriesSubspace&&) = default;
+
+  PoissonSeriesSubspace& operator=(PoissonSeriesSubspace const&) = default;
+  PoissonSeriesSubspace& operator=(PoissonSeriesSubspace&&) = default;
```
Comment (pleroy) [inline numerics/poisson_series_basis.hpp:31]: I suspect that these 4 declarations could be removed.

File: numerics/poisson_series_basis.hpp [hunk 5]
```diff
@@ -17,6 +17,41 @@ using geometry::Hilbert;
...
+
+  // Whether the subspaces |v| and |w| are orthogonal.
+  // TODO(egg): when we take parity into account, orthogonality will be defined
```
Comment (pleroy) [inline numerics/poisson_series_basis.hpp:34]: When

File: numerics/poisson_series_basis_body.hpp (hunk 1 above)
Comment (pleroy) [reply numerics/poisson_series_basis_body.hpp:180]: Same formatting comment, see line 169.

File: numerics/poisson_series_basis_body.hpp [hunk 6]
```diff
@@ -164,6 +174,20 @@ std::array<Series, sizeof...(indices)> AperiodicSeriesGenerator<
...
 
+template<typename Series, int degree, int dimension, std::size_t... indices>
+std::array<PoissonSeriesSubspace, sizeof...(indices)> AperiodicSeriesGenerator<
+    Series,
+    degree,
+    dimension,
+    std::index_sequence<indices...>>::Subspaces(Instant const& origin) {
+  return {PoissonSeriesSubspace{
+      static_cast<PoissonSeriesSubspace::Coordinate>(indices % dimension),
+      // The degree of the ith polynomial is i / dimension, so its parity is
```
Comment (pleroy) [inline numerics/poisson_series_basis_body.hpp:185]: There is no `i`, it's called `indices`.

File: numerics/poisson_series_basis_body.hpp (hunk 2 above)
Comment (pleroy) [reply numerics/poisson_series_basis_body.hpp:230]: Same.

File: numerics/poisson_series_basis_body.hpp [hunk 7]
```diff
@@ -205,6 +251,13 @@ auto PoissonSeriesBasisGenerator<Series, degree>::Basis(Instant const& origin)
              BasisElements(origin);
 }
 
+template<typename Series, int degree>
+inline auto PoissonSeriesBasisGenerator<Series, degree>::Subspaces(
```
Comment (pleroy) [inline numerics/poisson_series_basis_body.hpp:255]: Remove `inline`.

File: numerics/poisson_series_basis_body.hpp [hunk 8]
```diff
@@ -213,6 +266,16 @@ auto PoissonSeriesBasisGenerator<Series, degree>::Basis(
              BasisElements(ω, origin);
 }
 
+template<typename Series, int degree>
+inline auto
```
Comment (pleroy) [inline numerics/poisson_series_basis_body.hpp:270]: Remove `inline`.

**Review by pleroy — APPROVED** [review]

**Comment by pleroy** [comment]
I can haz tests?

File: numerics/poisson_series_basis_body.hpp (hunk 6 above)
Comment (eggrobin) [reply numerics/poisson_series_basis_body.hpp:185]: Yes, but this sentence is illegible if expressed in terms of a pack.

**Pull request merged**